File: GFG/Hashing/longsbar.py
def longest_subarray_with_sum(arr, target_sum):
    sum_map = {}
    max_length = 0
    curr_sum = 0

    for i in range(len(arr)):
        curr_sum += arr[i]

        if curr_sum == target_sum:
            max_length = i + 1

        if curr_sum - target_sum in sum_map:
            max_length = max(max_length, i - sum_map[curr_sum - target_sum])

        if curr_sum not in sum_map:
            sum_map[curr_sum] = i

    return max_length


arr = [5,8,-4,-4,9,2,-2]
target_sum = 0
result = longest_subarray_with_sum(arr, target_sum)
print("Longest subarray length:", result)

File: GFG/Hashing/longsbar.cpp
#include <iostream>
#include <unordered_map>
using namespace std;

int longestSubarrayWithSum(int arr[], int n, int targetSum) {
    unordered_map<int, int> sumMap;
    int maxLength = 0;
    int currentSum = 0;

    for (int i = 0; i < n; i++) {
        currentSum += arr[i];

        if (currentSum == targetSum) {
            maxLength = i + 1;
        }

        if (sumMap.find(currentSum - targetSum) != sumMap.end()) {
            maxLength = max(maxLength, i - sumMap[currentSum - targetSum]);
        }

        if (sumMap.find(currentSum) == sumMap.end()) {
            sumMap[currentSum] = i;
        }
    }

    return maxLength;
}

int main() {
    int arr[] = {3, 1, 2, -2, 4, 2, 1};
    int n = sizeof(arr) / sizeof(arr[0]);
    int targetSum = 4;
    int result = longestSubarrayWithSum(arr, n, targetSum);
    cout << "Longest subarray length: " << result << endl;
    return 0;
}
